README.md
# Service Canada Labs / Alpha Site

Service Canada Labs is a product being built by the Digital Experience and Client Data (DECD) unit within Employment and
Social Development Canada (ESDC) which is a department of the Government of Canada. This product is part of a larger
project occurring within ESDC to modernize benefits delivery to Canadian citizens and residents aptly called Benefits
Delivery Modernization (BDM) . The goal of this product is to be a digital platform where the general public is able to
be see, test and provide feedback on projects at different stages of design and development by DECD and partners.
This will allow early and rapid collection of user feedback which is vital in delivering useful products with excellent
user experience.

## Getting Started

Run the development server:

```bash
yarn install
yarn dev
```

Create and run a production build locally\*\*\*:

\*\*\* Will require `AEM_GRAPHQL_ENDPOINT=https://www.canada.ca/content/_cq_graphql/decd-endc/endpoint.json` and `ADOBE_ANALYTICS_URL=` the second being avariable with no value. These env vars should be in a top-level .env file

```bash
yarn build
yarn start
```

Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

## Configuration

There are two required variables: `AEM_GRAPHQL_ENDPOINT=https://www.canada.ca/content/_cq_graphql/decd-endc/endpoint.json` and `ADOBE_ANALYTICS_URL=`

`AEM_GRAPHQL_ENDPOINT` is used to fetch the data for pages and components across the site.

## Enabling Report a Problem

Required environment variables:

```code
REPORT_A_PROBLEM_ENABLED=true
NOTIFY_BASE_API_URL=https://api.notification.canada.ca
NOTIFY_REPORT_A_PROBLEM_EMAIL=email which will receive the feedback from the form
NOTIFY_API_KEY=ask for this
NOTIFY_REPORT_A_PROBLEM_TEMPLATE_ID=ask for this
```

1. Ask the dev team for GCNotify API key and template keys

2. Start the application as usual

## Environment Variables

Here is a list of all the environment variables used to configure the application along with what they do.

`REPORT_A_PROBLEM_ENABLED`: Feature flag for the report a problem API to enable sending to GC Notify. Leave undefined if
you do not want to enable this.

`NEXT_PUBLIC_NOTIFY_REPORT_A_PROBLEM_TEMPLATE_ID`: The Notify template ID for the report a problem email template

`NEXT_PUBLIC_NOTIFY_REPORT_A_PROBLEM_EMAIL`: The email to send it to

`NOTIFY_API_KEY`: The [Notify API key](https://documentation.notification.canada.ca/en/start.html#headers)

`NOTIFY_BASE_API_URL`: The base url for the Notify api. This value should be https://api.notification.canada.ca

`ADOBE_ANALYTICS_URL`: URL for adobe analytics. It is found in the documentation for Adobe Analytics
installation.

## Checking Deployed Version
Run the following bash script to check whether the latest (on `main`) has been deployed to various environments:
```bash
$ sh ./check-version.sh
```

## Learn More About the Framework We're Using

To learn more about Next.js, take a look at the following resources:

- [Next.js Documentation](https://nextjs.org/docs) - learn about Next.js features and API.
- [Learn Next.js](https://nextjs.org/learn) - an interactive Next.js tutorial.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## Added

## Changed

## Fixed

## [v1.9.0] - 2023-07-25

## Added

- Benefits Navigator page
- Breadcrumb data for pages is passed through new util function to avoid hardcoding

## Changed

- Moved tag element in Card.js to the right of the title element
- Various fixes to address alignment with design

## Fixed

## [v1.8.0] - 2023-07-18

## Added

- Added OAS update page

## Changed

- Modified OAS overview page design
- Added update/blog cards to OAS overview page

## Fixed

## [v1.7.0] - 2023-07-14

## Added

- SCCH page
- Filter for projects with status of "current" on the home page

## Changed

- Card.js now uses object-contain for images

## Fixed

## [v1.6.1] - 2023-07-13

## Added

## Changed

## Fixed

- Fixed issue with ProjectInfo tooltip shifting content at some display sizes

## [v1.6.0] - 2023-07-12

## Added

- margin-bottom: 24px to headings h1 through h5

## Changed

- Change heading in home.js to use SCDS Heading component
- Deprecate use of Alert and replaced with SCDS ContextualAlert
- Repositioned homepage Alert
- Update font colour across the site to #333333
- Small changes to Card.js
- Increased h3 line-height
- Removed class overriding default border colour on Card.js
- Updated headings sections to match new design with image under main H1
- Modify font-sizes to align with SCDS responsive sizing

## Fixed

## [v1.5.2] - 2023-05-09

## Changed

- ProjectInfo.js now displays labels and their content in a single column
- ProjectInfo.js FocusTrap now behaves like an accordion, pushing proceeding content beneath it rather than displaying overtop of it
- Hid project pages without links on the main page
- Right-clicking anywhere on Card.js allows to open the link in a new tab
- Project pages now display a question mark for information instead of an "i"
- Signup pages return 404 in production
- Remove `<a>` tag nested in another `<a>` from Card.js

## [v1.5.1] - 2023-04-12

## Added

- Add dateModifiedOverride to oas-benefits-estimator/index.js

## [v1.5.0] - 2023-04-12

## Added

- Added a button to resend email after user signup
- Added OWASP ZAP testing to github action
- Added copied state for IE error
- Added `/notsupported` which IE users are redirected to by making use of the next.js redirects function
- Added the `CopyToClickboard` component as part of the `/notsupported` page
- Added IE specific styling to `globals.css` for the `/notsupported` page
- Added IE detection to `/404` and `/500` pages as they don't auto redirect using the next.js redirects function
- Added language detection to redirect user to the page in prefered language
- Added a signup info page prior to the signup page
- Added metadata for `/500`, `/error`, `/404`, `/notsupported`, `/projects/virtual-assistant`, `/projects/digital-centre`, `/signup-info`, and the splash page
- Added an optional `Are you a public servant` question to sign-up form
- Added an `Confirm email` field to sign-up form
- Added a wrapper class that uses singleton pattern to manage Strapi data
- Added `autoComplete` prop type to TextField component
- Added `imageSource` prop type to ActionButton component
- Added `imageAlt` prop type to ActionButton component
- Added `imageSpanText` prop type to ActionButton component
- Added `imageSpanClass` prop type to ActionButton component
- Added remaining character counter to feedback widget
- Added the `stripFeedback` helper function which strips PII from both the feedback and report a problem forms
- Added error label in report a problem widget
- Added environment variable for our project version to our Next.js config for metadata purposes
- Added image in Experiment card
- Added Sharp for image optimization in production
- Added the `Alert` component for contextual alerts
- Added the `TableOfContents` component to convert page headings into a table of contents
- Added the `default` locale and middleware to redirect all routes using this locale to their corresponding `en` routes
- Added `/signup-review` page where users can review their signup information before final submission for registration
- Added `check-email` API endpoint for checking if a user exists
- Added Update/Blog page for Virtual Assistant project updates
- Added /api/healthcheck endpoint

## Changed

- Add collape and expand state on the checkbox on the Report A Problem form
- Add border to feedback popup, and give background a darker color
- Lock scroll when feedback popup is active
- Deactivate auto focus when feedback popup is active
- Updated introduction paragraph on the project page
- Removed all the initial text on signup page
- Updated metadata on every page according to the master content inventory
- Updated project to use Tailwind V3
- Updated the language question on the signup page by putting `aria-required` on the legend and removing `required` from each radio button so screen readers only announce required once for the grouping
- Removed red `required` from signup page fields to match Figma design
- Updated feedback widget's color scheme to improve a11y
- Move feedback error message in the feedbackinfo div which links to the textarea
- Updated project and Dockerfile to use yarn instead of npm
- Updated signup error labels and remove optional on labels
- Update project page to read content from cms
- Update home page data model to read content from cms
- Update about page to read content from cms
- Changed number of steps for signup and unsubscribe from 3 to 2
- Removed srSpeak function from feedback widget and instead put `aria-live="polite"` on the form
- Update signup-info page to read content from cms
- Removed `experience` email from privacy page, Report a Problem and Feedback components
- Updated to Next.js `v12.1.6`

pages/404.js
import Head from "next/head";
import { useTranslation } from "next-i18next";
import { serverSideTranslations } from "next-i18next/serverSideTranslations";
import Link from "next/link";
import { ReportAProblem } from "../components/organisms/ReportAProblem";
import { ActionButton } from "../components/atoms/ActionButton";
import { useEffect, useState } from "react";
import { useRouter } from "next/router";
import aemServiceInstance from "../services/aemServiceInstance";

export default function error404(props) {
  const { t } = useTranslation("common");
  const [loaded, setLoaded] = useState(false);
  const router = useRouter();
  const [pageData] = useState(props.pageData.item);

  useEffect(() => {
    if (props.adobeAnalyticsUrl) {
      window.adobeDataLayer = window.adobeDataLayer || [];
      window.adobeDataLayer.push({ event: "pageLoad" });
    }

    //If using Internet Explorer, redirect to /notsupported
    if (window.document.documentMode) {
      router.push("/notsupported");
    } else {
      setLoaded(true);
    }
  }, []);

  //If using Internet Explorer, render empty page so page doesn't flash before redirect
  if (!loaded) {
    return <div></div>;
  }

  return (
    <>
      <div className="min-h-screen relative">
        <Head>
          {props.adobeAnalyticsUrl ? (
            <script src={props.adobeAnalyticsUrl} />
          ) : (
            ""
          )}

          {/* Primary HTML Meta Tags */}
          <title data-gc-analytics-error="404">
            {pageData.scContentEn.json[0].content[0].value} (404) |{" "}
            {pageData.scContentFr.json[0].content[0].value} (404)
          </title>
          <meta
            name="description"
            content={`${t("404errorMetaDescription")}`}
          />
          <meta name="author" content="Service Canada" />
          <link rel="icon" href="/favicon.ico" />
          <link rel="schema.dcterms" href="http://purl.org/dc/terms/" />

          {/* DCMI Meta Tags */}
          <meta name="dcterms.title" content={`404 — ${t("siteTitle")}`} />
          <meta
            name="dcterms.language"
            content={props.locale === "en" ? "eng" : "fra"}
            title="ISO639-2/T"
          />
          <meta
            name="dcterms.description"
            content={`${t("404errorMetaDescription")}`}
          />
          <meta
            name="dcterms.subject"
            title="gccore"
            content={t("metaSubject")}
          />
          <meta name="dcterms.creator" content="Service Canada" />
          <meta name="dcterms.accessRights" content="2" />
          <meta
            name="dcterms.service"
            content="ESDC-EDSC_SCLabs-LaboratoireSC"
          />
          <meta name="dcterms.issued" title="W3CDTF" content="2021-06-01" />
          <meta name="dcterms.modified" title="W3CDTF" content="2021-12-16" />
          <meta name="dcterms.spatial" content="Canada" />

          {/* Open Graph / Facebook */}
          <meta property="og:type" content="website" />
          <meta property="og:locale" content={props.locale} />
          <meta
            property="og:url"
            content={
              "https://alpha.service.canada.ca/" + `${props.locale}` + "/404"
            }
          />
          <meta property="og:title" content={`404 — ${t("siteTitle")}`} />
          <meta
            property="og:description"
            content={`${t("404errorMetaDescription")}`}
          />
          <meta property="og:image" content={`${t("metaImage")}`} />
          <meta property="og:image:alt" content={`${t("siteTitle")}`} />

          {/* Twitter */}
          <meta property="twitter:card" content="summary_large_image" />
          <meta
            property="twitter:url"
            content={
              "https://alpha.service.canada.ca/" + `${props.locale}` + "/404"
            }
          />
          <meta property="twitter:title" content={`500 — ${t("siteTitle")}`} />
          <meta name="twitter:creator" content="Service Canada" />
          <meta
            property="twitter:description"
            content={`${t("404errorMetaDescription")}`}
          />
          <meta property="twitter:image" content={`${t("metaImage")}`} />
          <meta property="twitter:image:alt" content={`${t("siteTitle")}`} />
        </Head>
        <section className="layout-container pb-44">
          <div className="pt-6">
            <img
              src={`https://www.canada.ca${
                props.locale === "en"
                  ? pageData.scGcImages[0].scImageEn._path
                  : pageData.scGcImages[0].scImageFr._path
              }`}
              alt={
                props.locale === "en"
                  ? pageData.scGcImages[0].scImageAltTextEn
                  : pageData.scGcImages[0].scImageAltTextFr
              }
              width={575}
              height={59}
            />
          </div>
          <div className="flex flex-col lg:flex-row justify-between items-center lg:items-start mt-8">
            <div>
              <div className="relative h-auto xl:w-96 xxl:w-400px lg:w-72 xl:h-400px lg:h-500px mb-8 lg:mb-0">
                <h1 className="font-bold font-display mb-4">
                  {pageData.scContentEn.json[0].content[0].value}
                </h1>
                <p className="font-bold font-body mb-8">
                  {pageData.scContentEn.json[1].content[0].value}
                </p>
                <p className="font-body text-sm mb-4 leading-30px">
                  {pageData.scContentEn.json[2].content[0].value}
                </p>
                <div className="flex">
                  <span className="error404-link" />
                  <p className="font-body text-sm leading-30px">
                    {pageData.scContentEn.json[3].content[0].value}
                    <Link
                      href={pageData.scContentEn.json[3].content[1].data.href}
                      className="underline hover:text-canada-footer-hover-font-blue text-canada-footer-font"
                    >
                      {pageData.scContentEn.json[3].content[1].value}
                    </Link>
                  </p>
                </div>
              </div>
              <ReportAProblem language={"en"} />
            </div>
            <div className="flex items-center justify-center circle-background my-8 mx-4 lg:mt-0 lightbulb-bg shrink-0">
              <span className="relative lightbulb">
                <img
                  src={`https://www.canada.ca${
                    props.locale === "en"
                      ? pageData.scImageList[0].scImageEn._path
                      : pageData.scImageList[0].scImageFr._path
                  }`}
                  alt=""
                />
              </span>
            </div>
            <div>
              <div
                className="relative h-auto xl:w-96 xxl:w-400px lg:w-72 xl:h-400px lg:h-500px mb-8 lg:mb-0"
                lang="fr"
              >
                <h1 className="font-bold font-display mb-4">
                  {pageData.scContentFr.json[0].content[0].value}
                </h1>
                <p className="font-bold font-body mb-8">
                  {pageData.scContentFr.json[1].content[0].value}
                </p>
                <p className="font-body text-sm mb-4 leading-30px">
                  {pageData.scContentFr.json[2].content[0].value}
                </p>
                <div className="flex">
                  <span className="error404-link" />
                  <p className="font-body text-sm leading-30px">
                    {pageData.scContentFr.json[3].content[0].value}
                    <Link
                      href={pageData.scContentFr.json[3].content[1].data.href}
                      className="underline hover:text-canada-footer-hover-font-blue text-canada-footer-font"
                    >
                      {pageData.scContentFr.json[3].content[1].value}
                    </Link>
                  </p>
                </div>
              </div>
              <ReportAProblem language="fr" />
            </div>
          </div>
        </section>
        <footer className="h-100px w-screen bg-footer-background-color absolute bottom-0">
          <div className="layout-container flex justify-between lg:flex-row-reverse pt-4 lg:pt-0 lg:mt-8">
            <ActionButton
              id="404TopOfPageButton"
              href="#"
              custom="text-left w-32 flex flex-col lg:hidden"
              text="Top of page / Haut de la page"
              icon="icon-up-caret"
              iconEnd
            />
            <span className="relative footer-logo">
              <img
                src={`https://www.canada.ca${
                  props.locale === "en"
                    ? pageData.scGcImages[1].scImageEn._path
                    : pageData.scGcImages[1].scImageFr._path
                }`}
                alt={
                  props.locale === "en"
                    ? pageData.scGcImages[1].scImageAltTextEn
                    : pageData.scGcImages[1].scImageAltTextFr
                }
              />
            </span>
          </div>
        </footer>
      </div>
      {props.adobeAnalyticsUrl ? (
        <script type="text/javascript">_satellite.pageBottom()</script>
      ) : (
        ""
      )}
    </>
  );
}

export const getStaticProps = async ({ locale }) => {
  const { data } = await aemServiceInstance.getFragment("error404Query");

  return {
    props: {
      locale: locale,
      adobeAnalyticsUrl: process.env.ADOBE_ANALYTICS_URL,
      ...(await serverSideTranslations("en", ["common"])),
      ...(await serverSideTranslations("fr", ["common"])),
      pageData: data.scLabsErrorPagev1ByPath,
    },
  };
};
